### PR TITLE
[MachineBlockPlacement][X86] Use max of MDAlign and TLIAlign to align Loops.

### PR DESCRIPTION
**Pull request by FreddyLeaf**
This patch added backend consumption on a new loop metadata:
!1 = !{!"llvm.loop.align", i32 64}
which is generated from clang's new loop attribute:
[[clang::code_align()]]
clang patch: #70762

### CODE REVIEW

**Comment by FreddyLeaf** [comment]
ping for review

**Comment by KanRobert** [comment]
> #70762

Should we wait for #70762 ?

File: llvm/include/llvm/CodeGen/MachineLoopInfo.h [hunk 1]
```diff
@@ -66,6 +67,14 @@ class MachineLoop : public LoopBase<MachineBasicBlock, MachineLoop> {
   /// it returns an unknown location.
   DebugLoc getStartLoc() const;
 
+  /// \brief Find the llvm.loop metadata for this loop.
+  /// If each branch to the header of this loop contains the same llvm.loop
+  /// metadata, then this metadata node is returned. Otherwise, if any
+  /// latch instruction does not contain the llvm.loop metadata or
+  /// multiple latch instructions contain different llvm.loop metadata nodes,
+  /// then null is returned.
```
Comment (KanRobert) [inline llvm/include/llvm/CodeGen/MachineLoopInfo.h:75]: Why null? Should we use the max value of alignment if multiple latch instructions contain different `llvm.loop` metadata nodes.

File: llvm/lib/CodeGen/MachineLoopInfo.cpp [hunk 2]
```diff
@@ -151,6 +151,54 @@ MachineLoopInfo::findLoopPreheader(MachineLoop *L, bool SpeculativePreheader,
...
+      return nullptr;
+    if (const auto *TI = BB->getTerminator())
+      LoopID = TI->getMetadata(LLVMContext::MD_loop);
+  } else if (auto *MBB = getHeader()) {
+    // There seem to be multiple latch blocks, so we have to
+    // visit all predecessors of the loop header and check
+    // their terminating instructions for the metadata.
+    if (const auto *H = MBB->getBasicBlock()) {
+      // Walk over all blocks in the loop.
+      for (auto *MBB : this->blocks()) {
```
Comment (phoebewang) [inline llvm/lib/CodeGen/MachineLoopInfo.cpp:170]: const?

**Comment by FreddyLeaf** [comment]
> > #70762
> 
> Should we wait for #70762 ?

I think we can do it parallel. And backend consumption can land first?

**Comment by FreddyLeaf** [comment]
CFE patch landed: https://github.com/llvm/llvm-project/pull/70762

File: llvm/lib/CodeGen/MachineLoopInfo.cpp (hunk 2 above)
Comment (FreddyLeaf) [reply llvm/lib/CodeGen/MachineLoopInfo.cpp]: Addressed in latest commit.

File: llvm/lib/CodeGen/MachineLoopInfo.cpp [hunk 3]
```diff
@@ -151,6 +151,54 @@ MachineLoopInfo::findLoopPreheader(MachineLoop *L, bool SpeculativePreheader,
...
+        const auto *BB = MBB->getBasicBlock();
+        if (!BB)
+          return nullptr;
+        const auto *TI = BB->getTerminator();
+        if (!TI)
+          return nullptr;
+        MDNode *MD = nullptr;
+        // Check if this terminating instruction jumps to the loop header.
+        for (const auto *S : successors(TI)) {
+          if (S == H) {
```
[the remark this answers is not in the document]
Comment (FreddyLeaf) [reply llvm/lib/CodeGen/MachineLoopInfo.cpp]: Addressed in latest commit.

File: llvm/include/llvm/CodeGen/MachineLoopInfo.h (hunk 1 above)
Comment (FreddyLeaf) [reply llvm/include/llvm/CodeGen/MachineLoopInfo.h:75]: it is allowed to not set this loop metadata node from front end and patch wants to set a default alignment for such situation.

File: llvm/lib/CodeGen/MachineBlockPlacement.cpp [hunk 4]
```diff
@@ -2919,8 +2919,30 @@ void MachineBlockPlacement::alignBlocks() {
     if (!L)
       continue;
 
-    const Align Align = TLI->getPrefLoopAlignment(L);
-    if (Align == 1)
+    const Align TLIAlign = TLI->getPrefLoopAlignment(L);
+    unsigned MDAlign = 1;
+    MDNode *LoopID = L->getLoopID();
+    if (LoopID) {
+      for (unsigned i = 1, e = LoopID->getNumOperands(); i < e; ++i) {
```
[the remark this answers is not in the document]
Comment (FreddyLeaf) [reply llvm/lib/CodeGen/MachineBlockPlacement.cpp]: eb597889e79b6d8cf113d5d81f2c38b9b5a203da

File: llvm/include/llvm/CodeGen/MachineLoopInfo.h [hunk 5]
```diff
@@ -66,6 +67,14 @@ class MachineLoop : public LoopBase<MachineBasicBlock, MachineLoop> {
   /// it returns an unknown location.
   DebugLoc getStartLoc() const;
 
+  /// \brief Find the llvm.loop metadata for this loop.
+  /// If each branch to the header of this loop contains the same llvm.loop
```
Comment (KanRobert) [inline llvm/include/llvm/CodeGen/MachineLoopInfo.h:71]: Should we add a blank line after line 70?
https://llvm.org/docs/CodingStandards.html

> The first sentence (or a passage beginning with \brief) is used as an abstract. Any additional information should be separated by a blank line. If an algorithm is based on a paper or is described in another source, provide a reference.

Comment (FreddyLeaf) [reply llvm/include/llvm/CodeGen/MachineLoopInfo.h:71]: Will remove `\brief`.

Comment (FreddyLeaf) [reply llvm/include/llvm/CodeGen/MachineLoopInfo.h:71]: 830699979b3d24edd97ddbb389f348491106776c

**Review by phoebewang — APPROVED** [review]
LGTM.

**Review by KanRobert — APPROVED** [review]
LGTM

File: llvm/lib/CodeGen/MachineLoopInfo.cpp [hunk 6]
```diff
@@ -151,6 +151,53 @@ MachineLoopInfo::findLoopPreheader(MachineLoop *L, bool SpeculativePreheader,
   return Preheader;
 }
 
+MDNode *MachineLoop::getLoopID() const {
+  MDNode *LoopID = nullptr;
+  if (const auto *MBB = findLoopControlBlock()) {
```
[the remark this answers is not in the document]
Comment (FreddyLeaf) [reply llvm/lib/CodeGen/MachineLoopInfo.cpp:156]: From my understanding, if the Loop has one existing BB, the loop metadata is supposed to be attached there?

[the remark this answers is not in the document]
Comment (kortbeek-snps) [reply llvm/lib/CodeGen/MachineLoopInfo.cpp:156]: Thanks for the quick response, @FreddyLeaf!

I don't see this reflected in the `Loop::setLoopID` function in the middle-end [llvm/lib/Analysis/LoopInfo.cpp#L526](https://github.com/llvm/llvm-project/blob/8a53324aa5c98787fe2334180a26b426f793d139/llvm/lib/Analysis/LoopInfo.cpp#L526), and the `llvm.loop` documentation also only mentions that the loop metadata is connected to the branch instruction in the loop latch (https://llvm.org/docs/LangRef.html#llvm-loop). Is this exception documented somewhere?

[the remark this answers is not in the document]
Comment (FreddyLeaf) [reply llvm/lib/CodeGen/MachineLoopInfo.cpp:156]: Emm, looks like my understanding is outdated. I agree to mirror the middle end's implementation. Would you help do a refinement on this?

**Pull request merged**